src/lib.rs

```diff
@@ -0,0 +1,450 @@
+        body.blocks.push(make_block(block_el));
+        let next_block = get_next_block_element(block_el);
+        match next_block {
+            Some(next_block_el) => {
+                block_el = next_block_el;
+            },
+            None => {
+                break;
+            }
+        }
+    }
+
+    body
+}
+
+fn make_block(block_el: Element) -> Block {
+    let mut block = Block::new();
+    for attribute in block_el.attributes().iter() {
+        let name = attribute.name().local_part();
+        let value = attribute.value().to_string();
+        match name {
+            "type" => { block.block_type = value; },
+            "id" => { block.id = value; },
+            _ => {}
+        }
+    }
+    for child in block_el.children().iter() {
+        match child {
+            &ChildOfElement::Element(child_el) => {
+                let child_name = child_el.name().local_part();
+                match child_name {
+                    "statement" => {
+                        let statement_el = child_el;
+                        let statement_name = get_attribute(statement_el, "name").unwrap();
+                        let statement_body = match get_first_child_element(statement_el) {
+                            Some(first_child_block) => make_statement_body(first_child_block),
+                            None => StatementBody::new()
+                        };
+                        block.add_statement_body(statement_name, statement_body);
+                    },
+                    "field" => {
+                        let field_el = child_el;
+                        let field_name = get_attribute(field_el, "name").unwrap();
+                        let field_value = make_field_value(field_el);
+                        block.add_field(field_name, field_value);
+                    },
+                    _ => {}
+                }
+            },
+            _ => {}
+        }
+    }
+    block
+}
+
+fn make_field_value(field_el: Element) -> FieldValue {
+    for child in field_el.children().iter() {
+        match child {
+            &ChildOfElement::Text(text_node) => {
+                let value = text_node.text().to_string();
+                return FieldValue::SimpleField(value);
+            },
+            _ => panic!("TODO: Implement expression fields")
+        }
+    }
+    panic!("Expected child nodes for field");
+}
+
+// General DOM utilities
+
+fn get_xml_element(document: Document) -> Element {
+    let root: Root = document.root();
+    let root_children = root.children();
+    for child in root_children.iter() {
+        match child {
+            &ChildOfRoot::Element(el) => {
+                let element_name = el.name().local_part();
+                if element_name == "xml" {
+                    return el;
+                }
+            },
+            _ => {}
+        }
+    }
+    panic!("Cannot find xml element!");
+}
+
+fn get_first_child_element(element: Element) -> Option<Element> {
```

Comment:
Just as an example, for a function signature like this, we might have the following documentation:

> This function *consumes* an element, and returns its first child.

The reason we say **consume** is because at the end of this function all the memory that was allocated for the passed in `Element` was freed. We **moved** the first child _out of_ `element` so that it would not be consumed when this process occurs. If you wanted to use `element` later (like if this function was just `peek_at_first_child`) then you would not want to take ownership of `element`, but simply **borrow** it (`&Element`).

Reply:
That's a good point. I don't really want to consume the element here so I should be borrowing it.